File: packages/kotlin/cnos-kotlin/src/main/kotlin/ai/kitsy/cnos/internal/Manifest.kt
package ai.kitsy.cnos.internal

import ai.kitsy.cnos.ServerProjection
import ai.kitsy.cnos.VaultDefinition

internal data class NamespaceDef(
    val kind: String,
    val runtime: Boolean,
    val shareable: Boolean,
    val sensitive: Boolean,
    val prefix: String?
)

internal data class BootstrappedManifest(
    val namespaces: Map<String, NamespaceDef>,
    val frameworks: Map<String, String>,
    val envMappingExplicit: Map<String, String>,
    val vaults: Map<String, VaultDefinition>
) {
    fun getNamespaceDef(ns: String): NamespaceDef =
        namespaces[ns] ?: NamespaceDef(KIND_DATA, false, true, false, null)

    companion object {
        const val KIND_DATA = "data"

        val DEFAULT_NAMESPACES: Map<String, NamespaceDef> = mapOf(
            "value"  to NamespaceDef(KIND_DATA, false, true,  false, null),
            "secret" to NamespaceDef(KIND_DATA, false, false, true,  null),
            "meta"   to NamespaceDef(KIND_DATA, false, false, false, null),
            "public" to NamespaceDef(KIND_DATA, false, true,  false, null),
            "process" to NamespaceDef(KIND_DATA, true,  false, false, null)
        )

        val DEFAULT_FRAMEWORKS: Map<String, String> = mapOf(
            "next"       to "NEXT_PUBLIC_",
            "vite"       to "VITE_",
            "react"      to "REACT_APP_",
            "gatsby"     to "GATSBY_",
            "expo"       to "EXPO_PUBLIC_",
            "nuxt"       to "NUXT_PUBLIC_",
            "svelte"     to "PUBLIC_",
            "astro"      to "PUBLIC_",
            "angular"    to "NG_APP_",
            "webpack"    to "PUBLIC_"
        )

        fun fromProjection(projection: ServerProjection): BootstrappedManifest {
            val ns = DEFAULT_NAMESPACES.toMutableMap()
            projection.meta.namespaces?.forEach { name ->
                ns.putIfAbsent(name, NamespaceDef(KIND_DATA, false, true, false, null))
            }
            return BootstrappedManifest(ns, DEFAULT_FRAMEWORKS.toMap(), emptyMap(), projection.vaults)
        }
    }
}

File: packages/rust/cnos/src/manifest.rs
use std::collections::HashMap;
use crate::projection::{ServerProjection, VaultDef};

#[derive(Debug, Clone)]
pub struct NamespaceDef {
    pub kind: String,
    pub shareable: bool,
    pub sensitive: bool,
    pub readonly: bool,
}

#[derive(Debug, Clone)]
pub struct RuntimeNamespaceDef {
    pub server_only: bool,
    pub built_in: bool,
}

#[derive(Debug, Clone)]
pub struct BootstrappedManifest {
    pub namespaces: HashMap<String, NamespaceDef>,
    pub runtime_namespaces: HashMap<String, RuntimeNamespaceDef>,
    pub frameworks: HashMap<String, String>,
    pub vaults: HashMap<String, VaultDef>,
}

pub fn default_namespace_defs() -> HashMap<String, NamespaceDef> {
    let mut m = HashMap::new();
    m.insert("value".into(), NamespaceDef { kind: "data".into(), shareable: true, sensitive: false, readonly: false });
    m.insert("secret".into(), NamespaceDef { kind: "data".into(), shareable: false, sensitive: true, readonly: false });
    m.insert("meta".into(), NamespaceDef { kind: "system".into(), shareable: false, sensitive: false, readonly: true });
    m.insert("process".into(), NamespaceDef { kind: "system".into(), shareable: false, sensitive: false, readonly: true });
    m.insert("public".into(), NamespaceDef { kind: "projection".into(), shareable: true, sensitive: false, readonly: true });
    m.insert("env".into(), NamespaceDef { kind: "projection".into(), shareable: true, sensitive: false, readonly: true });
    m
}

pub fn default_frameworks() -> HashMap<String, String> {
    let mut m = HashMap::new();
    m.insert("next".into(), "NEXT_PUBLIC_".into());
    m.insert("vite".into(), "VITE_".into());
    m.insert("nuxt".into(), "NUXT_PUBLIC_".into());
    m.insert("webpack".into(), "".into());
    m
}

pub fn bootstrapped_manifest_from_projection(projection: &ServerProjection) -> BootstrappedManifest {
    let mut namespaces = default_namespace_defs();
    for ns in &projection.meta.namespaces {
        namespaces.entry(ns.clone()).or_insert_with(|| NamespaceDef {
            kind: "data".into(), shareable: false, sensitive: false, readonly: false,
        });
    }

    let mut runtime_namespaces = HashMap::new();
    runtime_namespaces.insert("process".into(), RuntimeNamespaceDef { server_only: true, built_in: true });
    for ns in &projection.runtime_namespaces {
        if ns != "process" {
            runtime_namespaces.insert(ns.clone(), RuntimeNamespaceDef { server_only: true, built_in: false });
        }
    }

    BootstrappedManifest {
        namespaces,
        runtime_namespaces,
        frameworks: default_frameworks(),
        vaults: projection.vaults.clone(),
    }
}
